[Very very unfinished] OpenGL rendering pipeline

dpogue opened the pull request:
This allows running with the OpenGL pipeline on Windows by running with the `/Renderer=gl` command-line flag.

There is a lot of stuff broken, or missing, or causing crashes! This is definitely not ready for actual use, but I do want to start a conversation about how to keep this more closely in line with the work on master so I can avoid future rebasing headaches after major refactorings.

Note: All the X11/Linux client stuff has been stripped out here, this is just the OpenGL rendering pipeline.

dpogue commented:
This is currently blocked on #1269

dpogue commented:
Need to sort out an initialization order problem between `hsG3DDeviceSelector::sEnumerators` and `plGLEnumerate::enumerator` that is causing segfaults for people

Hoikas commented on Sources/Plasma/FeatureLib/pfDXPipeline/plDXDevice.cpp:
```diff
@@ -161,6 +183,257 @@ void plDXDevice::SetViewport()
+ */
+static uint32_t GetBufferFormatSize(uint8_t format)
+{
+    uint32_t  size = sizeof(float) * 6 + sizeof(uint32_t) * 2; // Position and normal, and two packed colors
```
```suggestion
    constexpr uint32_t size = sizeof(float) * 6 + sizeof(uint32_t) * 2; // Position and normal, and two packed colors
```

dpogue replied:
This can't be `constexpr` because we add to the `size` variable and then return it

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp:
```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+#ifdef USE_EGL
+#include <epoxy/egl.h>
+
+void InitEGLDevice(plGLDevice* dev)
```
```suggestion
static void InitEGLDevice(plGLDevice* dev)
```

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp:
```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+#include "hsWindows.h"
+#include <epoxy/wgl.h>
+
+void InitWGLDevice(plGLDevice* dev)
```
```suggestion
static void InitWGLDevice(plGLDevice* dev)
```

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp:
```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+    HGLRC ctx = nullptr;
+
+    do {
+        PIXELFORMATDESCRIPTOR pfd = {};
```
```suggestion
        PIXELFORMATDESCRIPTOR pfd{};
```

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp:
```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+    }
+}
+
+void FiniWGLDevice(plGLDevice* dev)
```
```suggestion
static void FiniWGLDevice(plGLDevice* dev)
```

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp:
```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+#ifdef HS_BUILD_FOR_MACOS
+#include <AvailabilityMacros.h>
+#include <OpenGL/OpenGL.h>
+void InitCGLDevice(plGLDevice* dev)
```
```suggestion
static void InitCGLDevice(plGLDevice* dev)
```

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp:
```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+    IGNORE_WARNINGS_END
+}
+
+void FiniCGLDevice(plGLDevice* dev)
```
```suggestion
static void FiniCGLDevice(plGLDevice* dev)
```

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp:
```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+            fErrorMsg = "Failed to attach EGL context to surface";
+            return false;
+        }
+    } //else
```
```suggestion
    }
```

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp:
```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+
+static uint32_t  IGetBufferFormatSize(uint8_t format)
+{
+    uint32_t  size = sizeof( float ) * 6 + sizeof( uint32_t ) * 2; // Position and normal, and two packed colors
```
```suggestion
    constexpr uint32_t  size = sizeof(float) * 6 + sizeof(uint32_t) * 2; // Position and normal, and two packed colors
```

dpogue replied:
This can't be `constexpr` because we add to the `size` variable and then return it

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDeviceRef.h:
```diff
@@ -47,11 +47,42 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 
 #include <epoxy/gl.h>
 
+inline int plGLVersion() {
```
```suggestion
inline int plGLVersion()
{
```

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLPipeline.cpp:
```diff
@@ -216,4 +680,1359 @@ void plGLPipeline::ResetDisplayDevice(int Width, int Height, int ColorDepth, boo
+{
+    plProfile_BeginTiming(RenderBuff);
+
+    typename DeviceType::VertexBufferRef* vRef = (typename DeviceType::VertexBufferRef*)vb;
+    typename DeviceType::IndexBufferRef* iRef = (typename DeviceType::IndexBufferRef*)ib;
```
What's with all of the `typename` stuff?

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLPipeline.cpp:
```diff
@@ -216,4 +680,1359 @@ void plGLPipeline::ResetDisplayDevice(int Width, int Height, int ColorDepth, boo
+
+        vec_dst[0] += (srcX * m00 + srcY * m01 + srcZ * m02) * wgt;
+        vec_dst[1] += (srcX * m10 + srcY * m11 + srcZ * m12) * wgt;
+        vec_dst[1] += (srcX * m20 + srcY * m21 + srcZ * m22) * wgt;
```
```suggestion
        vec_dst[2] += (srcX * m20 + srcY * m21 + srcZ * m22) * wgt;
```

See #1336

Hoikas commented on Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h:
```diff
@@ -957,6 +1096,137 @@ void pl3DPipeline<DeviceType>::Draw(plDrawable* d)
 }
 
 
+template <class DeviceType>
```
```suggestion
template<class DeviceType>
```

Hoikas commented on Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h:
```diff
@@ -957,6 +1096,137 @@ void pl3DPipeline<DeviceType>::Draw(plDrawable* d)
+        fDevice.FillIndexBufferRef(iRef, owner, idx);
+}
+
+template <class DeviceType>
```
```suggestion
template<class DeviceType>
```

Hoikas commented on Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h:
```diff
@@ -957,6 +1096,137 @@ void pl3DPipeline<DeviceType>::Draw(plDrawable* d)
+    }
+}
+
+template <class DeviceType>
```
```suggestion
template<class DeviceType>
```

Hoikas commented on Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h:
```diff
@@ -1459,6 +1729,17 @@ void pl3DPipeline<DeviceType>::ISetShadowFromGroup(plDrawableSpans* drawable, co
 }
 
 
+template <class DeviceType>
```
```suggestion
template<class DeviceType>
```

Hoikas commented on Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h:
```diff
@@ -1561,6 +1842,78 @@ void pl3DPipeline<DeviceType>::IFreeAvRT(plRenderTarget* tex)
 }
 
 
+template <class DeviceType>
```
```suggestion
template<class DeviceType>
```

Hoikas commented on Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h:
```diff
@@ -1561,6 +1842,78 @@ void pl3DPipeline<DeviceType>::IFreeAvRT(plRenderTarget* tex)
+}
+
+
+template <class DeviceType>
```
```suggestion
template<class DeviceType>
```

Hoikas commented on Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h:
```diff
@@ -1561,6 +1842,78 @@ void pl3DPipeline<DeviceType>::IFreeAvRT(plRenderTarget* tex)
+}
+
+
+template <class DeviceType>
```
```suggestion
template<class DeviceType>
```

Hoikas commented on Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h:
```diff
@@ -1561,6 +1842,78 @@ void pl3DPipeline<DeviceType>::IFreeAvRT(plRenderTarget* tex)
+}
+
+
+template <class DeviceType>
```
```suggestion
template<class DeviceType>
```

Hoikas commented on Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h:
```diff
@@ -1883,4 +2236,136 @@ void pl3DPipeline<DeviceType>::ILocalToWorldToDevice()
+    }
+};
+
+template <class DeviceType>
```
```suggestion
template<class DeviceType>
```

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLMaterialShaderRef.cpp:
```diff
@@ -0,0 +1,634 @@
+    this->uMatDiffuseCol  = glGetUniformLocation(fRef, "uDiffuseCol");
+    this->uMatDiffuseSrc  = glGetUniformLocation(fRef, "uDiffuseSrc");
+    this->uMatEmissiveCol = glGetUniformLocation(fRef, "uEmissiveCol");
+    this->uMatEmissiveSrc = glGetUniformLocation(fRef, "uEmissiveSrc");
+    this->uMatSpecularCol = glGetUniformLocation(fRef, "uSpecularCol");
+    this->uMatSpecularSrc = glGetUniformLocation(fRef, "uSpecularSrc");
+
+    // Fog inputs
+    this->uFogExponential = glGetUniformLocation(fRef, "uFogExponential");
+    this->uFogValues      = glGetUniformLocation(fRef, "uFogValues");
+    this->uFogColor       = glGetUniformLocation(fRef, "uFogColor");
+
+    // Lamp inputs
+    for (size_t i = 0; i < 8; i++) {
+        this->uLampSources[i].position     = glGetUniformLocation(fRef, ST::format("uLampSources[{}].position", i).c_str());
+        this->uLampSources[i].ambient      = glGetUniformLocation(fRef, ST::format("uLampSources[{}].ambient", i).c_str());
+        this->uLampSources[i].diffuse      = glGetUniformLocation(fRef, ST::format("uLampSources[{}].diffuse", i).c_str());
+        this->uLampSources[i].specular     = glGetUniformLocation(fRef, ST::format("uLampSources[{}].specular", i).c_str());
+        this->uLampSources[i].direction    = glGetUniformLocation(fRef, ST::format("uLampSources[{}].direction", i).c_str());
+        this->uLampSources[i].spotProps    = glGetUniformLocation(fRef, ST::format("uLampSources[{}].spotProps", i).c_str());
+        this->uLampSources[i].constAtten   = glGetUniformLocation(fRef, ST::format("uLampSources[{}].constAtten", i).c_str());
+        this->uLampSources[i].linAtten     = glGetUniformLocation(fRef, ST::format("uLampSources[{}].linAtten", i).c_str());
+        this->uLampSources[i].quadAtten    = glGetUniformLocation(fRef, ST::format("uLampSources[{}].quadAtten", i).c_str());
+        this->uLampSources[i].scale        = glGetUniformLocation(fRef, ST::format("uLampSources[{}].scale", i).c_str());
+    }
+
+    size_t layerCount = fMaterial->GetNumLayers();
+
+    this->uLayerMat.assign(layerCount, -1);
+    for (size_t i = 0; i < layerCount; i++) {
+        ST::string name = ST::format("uLayerMat{}", i);
+        this->uLayerMat[i] = glGetUniformLocation(fRef, name.c_str());
+    }
+
+    this->uTexture.assign(layerCount, -1);
+    for (size_t i = 0; i < layerCount; i++) {
+        ST::string name = ST::format("uTexture{}", i);
+        this->uTexture[i] = glGetUniformLocation(fRef, name.c_str());
+    }
+}
```
What's with all the explicit `this->` stuff?

Hoikas commented on Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp:
```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+
+bool plGLDevice::InitDevice()
+{
+#ifdef USE_EGL
+    // The USE_EGL define tells us whether the epoxy library includes support
+    // for attempting to use EGL on the current platform, but we still need to
+    // check if EGL is actually available at runtime.
+    //
+    // On Windows, this may be true in cases like the PowerVR SDK or when using
+    // ANGLE.
+    //
+    // On Linux, this should be true with mesa or nvidia drivers.
+    if (epoxy_has_egl() && fContextType == kNone)
+        InitEGLDevice(this);
+#endif
+
+#ifdef HS_BUILD_FOR_WIN32
+    if (fContextType == kNone)
+        InitWGLDevice(this);
+#endif
+
+#ifdef HS_BUILD_FOR_MACOS
+    if (fContextType == kNone)
+        InitCGLDevice(this);
+#endif
```
This really feels like a good opportunity for some virtual functions for each context type.

Hoikas replied:
This really feels like a good opportunity for some virtual functions for each context type.

dpogue commented:
I think I've addressed all the comment from the first round of feedback on this, I'm sure there are more items to address but I also recognize that it's daunting to review this
